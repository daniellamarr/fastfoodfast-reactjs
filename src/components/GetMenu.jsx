/* eslint-disable radix */
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getAllMenu } from '../actions/Menu';
import { addToCart } from '../actions/cartActions';

/**
 * GetMenu
 */
class GetMenu extends React.Component {
  /**
   * Constructor
   */
  constructor() {
    super();

    this.state = {
      quantity: 1,
    };

    this.onChange = this.onChange.bind(this);
  }

  /**
   * onChange
   * @param {object} e
   * @returns {null} null
   */
  onChange(e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  /**
   * Initialized functions for component
   * @returns {null} null
   */
  componentDidMount() {
    this.props.getAllMenu();
  }

  /**
   * @returns {JSX} jsx
   */
  render() {
    const menuItems = this.props.menuItems.map(menuItem => (
      <div key={menuItem.id} className="item">
        <img src={menuItem.image} alt="" />
        <div className="item-text">
          <a href={`/menu/${menuItem.id}`}>
            {menuItem.title}
            <h5 className="price">N{menuItem.price}</h5>
          </a>
          <div className="price-tag">
            <input
              type="number"
              id={`qty${menuItem.id}`}
              name='quantity'
              className="quantityField"
              defaultValue="1"
              min={1}
              max={10}
              onChange={this.onChange}
            />
            <br /><br />
            <button
              className="addtocart"
              onClick={() => this.props.addToCart({
                id: menuItem.id,
                title: menuItem.title,
                quantity: parseInt(this.state.quantity),
                price: menuItem.price,
                image: menuItem.image,
              })}
            >
              Add to cart
            </button>
          </div>
        </div>
      </div>
    ));
    return (
      <div>
        {this.props.menuStatus
          ? menuItems
          : <center>
            <img src="/images/loader.gif" alt="loading..." width="100px" />
          </center>
        }
        <div className="clear"></div>
      </div>
    );
  }
}

GetMenu.propTypes = {
  getAllMenu: PropTypes.func.isRequired,
  addToCart: PropTypes.func.isRequired,
  menuItems: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    price: PropTypes.string.isRequired,
  })),
  menuStatus: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  menuItems: state.menus.menus,
  menuStatus: state.menus.menuStatus,
});

export default connect(mapStateToProps, { getAllMenu, addToCart })(GetMenu);
