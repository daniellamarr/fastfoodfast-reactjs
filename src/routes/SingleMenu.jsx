/* eslint-disable radix */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getSingleMenu } from '../actions/Menu';
import GetMenu from '../components/GetMenu.jsx';
import { addToCart } from '../actions/cartActions';

/**
 * SingleMenu
 */
export class SingleMenu extends Component {
  /**
   * Constructor
   */
  constructor() {
    super();

    this.state = {
      quantity: 1,
    };

    this.onChange = this.onChange.bind(this);
  }

  /**
   * onChange
   * @param {object} e
   * @returns {null} null
   */
  onChange(e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  /**
   * @returns {null} null
   */
  componentDidMount() {
    const { id } = this.props.match.params;
    this.props.getSingleMenu(id);
  }

  /**
   * @returns {JSX} jsx
   */
  render() {
    return (
      <div>
        {
          this.props.menuStatus
        }
        <section
          id="menudetails"
          style={{
            backgroundImage: `url(${this.props.menu.image_path})`,
          }}
        >
          <div className="title">
            <h1>
              {this.props.menu.title}
            </h1>
            <h2>
              N{this.props.menu.price}
            </h2>
            <p>
            {this.props.menu.quantity} item(s) remaining in stock
            </p>
            <input
              type="number"
              id={`qty${this.props.menu.id}`}
              name='quantity'
              className="quantityField"
              defaultValue="1"
              min={1}
              max={10}
              onChange={this.onChange}
            />
            <button
              onClick={() => this.props.addToCart({
                id: this.props.menu.id,
                title: this.props.menu.title,
                quantity: parseInt(this.state.quantity),
                price: this.props.menu.price,
                image: this.props.menu.image_path,
              })}
            >
              Add to cart
            </button>
          </div>
        </section>
        <section id="foods">
          <div className="title">
            <h2>More varieties</h2><hr className="div" />
          </div>
          <GetMenu />
        </section>
      </div>
    );
  }
}

SingleMenu.propTypes = {
  getSingleMenu: PropTypes.func.isRequired,
  addToCart: PropTypes.func.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired,
    }),
  }),
  menu: PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    price: PropTypes.string,
    quantity: PropTypes.string,
    image_path: PropTypes.string,
  }),
  menuStatus: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  menu: state.menus.menu,
  menuStatus: state.menus.menuStatus,
});

export default connect(mapStateToProps,
  { getSingleMenu, addToCart })(SingleMenu);
