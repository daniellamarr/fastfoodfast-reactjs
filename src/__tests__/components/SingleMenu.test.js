/* eslint-disable no-undef */
import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { SingleMenu } from '../../routes/SingleMenu.jsx';
import { mapStateToProps } from '../../components/GetMenu.jsx';

const initialState = {
  menus: {
    menus: [
      {
        id: 1,
        title: 'test',
        price: '1000',
      },
    ],
    menuStatus: false,
  },
};

const middlewares = [thunk];

const mockStore = configureMockStore(middlewares);
const store = mockStore(initialState);

let wrapper;

const props = {
  getSingleMenu: () => {},
  addToCart: () => {},
  match: {
    params: {
      id: '1',
    },
  },
  menu: {
    id: 1,
    title: 'test menu',
    quantity: '30',
    price: '1000',
    image: 'image url',
  },
  menus: {
    menus: [{}],
  },
  menuStatus: true,
};

beforeEach(() => {
  wrapper = mount(
    <Provider store={store}>
      <SingleMenu {...props} />
    </Provider>,
  );
});

describe('SingleMenu Component', () => {
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('click addtocart button', () => {
    wrapper.find('button').simulate('click');
  });

  it('should simulate quantity text field change', () => {
    wrapper.find('.quantityField').simulate('change');
    // expect(wrapper.state().quantity).toEqual('');
  });

  it('should show previously rolled value', () => {
    expect(mapStateToProps(initialState).menuItems)
      .toEqual(initialState.menus.menus);
  });
});
