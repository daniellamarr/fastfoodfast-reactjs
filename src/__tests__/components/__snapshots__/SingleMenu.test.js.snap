// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleMenu Component should match snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeState": Object {},
    }
  }
>
  <SingleMenu
    addToCart={[Function]}
    getSingleMenu={[Function]}
    match={
      Object {
        "params": Object {
          "id": "1",
        },
      }
    }
    menu={Object {}}
    menuStatus={false}
    menus={
      Array [
        Object {},
      ]
    }
  />
</ContextProvider>
`;
